msgs/msg/SolutionInfo.msg

```diff
@@ -12,3 +12,7 @@ uint32 stage_id
 
 # markers, e.g. providing additional hints or illustrating failure
 visualization_msgs/Marker[] markers
+
+# user-defined values that can provide additional information,
+# e.g., whether to switch a suction gripper on or off or when to call some other execution interface
+Property[] external
```

Comment:
The description and the structure looks good to me, I'm only thinking about the name, as mentioned in the other thread. For the examples we discussed, this might be a good one though.

Reply:
I don't like the name `external` as well. Here, I would just name them `properties` as well.
```suggestion
Property[] properties
```

Reply:
If we go for the generic name, we will also include all INTERFACE-configured stage properties & interface state properties used for computing the solution.
These are currently unobservable from the outside.

I'm looking into these aspects right now anyway, so if you agree @rhaschke I can prepare a first patch these days.

Reply:
I don't see why the generic name implies that all the internal properties should be exposed as well.
Internally (in the MTC source code), I'm fine to name those properties `external_props`.

> I'm looking into these aspects right now anyway, so if you agree @rhaschke I can prepare a first patch these days.

Yes, in general, I like the idea of exported solution properties - just to introduce yet another name suggestion :wink:

Reply:
> I don't see why the generic name implies that all the internal properties should be exposed as well.

It's a matter of perspective of course, but for me "the properties of the solution" clearly refers to, e.g., the approach angle used to compute the particular grasp candidate and not so much to the question whether or not a suction gripper should be switched on in this solution.

This is why I proposed the name `external`. I would be fine with mixing these two concepts though if we introduce a single "solution properties" list.

Reply:
I think we agree that the semantics of the properties exposed in the message is fully up to the user. Hence, arguing with semantics is not meaningful here.
I see two types of solution properties (from an abstract level):
- properties required during computation
- properties exposed to other processes, e.g. execution via the solution message

Of course, one could publish both types of properties via this interface, but I don't see a reason to do so.
How we name those two types of properties is a different story. We should first decide on the intended semantics.
As the user (from the external view) only sees the exposed properties, I would name them in the message as `properties`.

Reply:
I mostly agree with your assessment and wholeheartedly disagree with your conclusion. :smile: 

> I think we agree that the semantics of the properties exposed in the message is fully up to the user.

That was the original idea for this thread. However, after you started arguing for the generic name I noticed the missing message interface to expose INTERFACE properties (and other internal variables the stage might consider as properties).
All static properties are of course part of the `description` topic, but these properties obviously can't be specified there.
The properties list in the panel lists empty entries for dynamic properties, even when you browse through the individual solutions. From a GUI perspective it would help a lot if these missing entries are read from the `Solution`.
So if we introduce a generic `properties` field here, I believe it should definitely include them.

At the same time, having two separate `Properties[]` fields in the `SolutionInfo` seems overly redundant and thus I'm open to merge the two concepts.
As a consequence, users can simply set own properties in a stage to expose them with the solution which avoids the need for another interface.

Reply:
I'm not sure what you mean by "interface properties", but from a user perspective, including the intended usage in the field's name would be helpful, even if it does not cover every use case. I was thinking of something specific rather than general, but `exported_properties` might not be bad.

Reply:
> So if we introduce a generic properties field here, I believe it should definitely include them.

Ok, I agree that publishing those properties might be useful for debugging and thus we should publish both types of properties. If we do so, a generic name like `properties` is perfect.